pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,460 @@
+        where the Lipschitz constant is given by :math:`\sum |c_i|` for Hamiltonian
+        coefficients :math:`c_i`.
+
+        Args:
+            coeffs (Sequence[float]): the coefficients of the terms in the Hamiltonian
+
+        Raises:
+            ValueError: if the learning rate is large than :math:`2/\sum |c_i|`
+        """
+        self.lipschitz = np.sum(np.abs(coeffs))
+
+        if self._stepsize > 2 / self.lipschitz:
+            raise ValueError(f"The learning rate must be less than {2 / self.lipschitz}")
+
+    def _single_shot_expval_gradients(self, expval_cost, args, kwargs):
+        """Compute the single shot gradients of an ExpvalCost object"""
+
+        qnodes = expval_cost.qnodes
+        coeffs = expval_cost.hamiltonian.coeffs
+        device = qnodes[0].device
+
+        self.check_device(device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(coeffs)
+
+        try:
+            if self.wrs:
+                grads = self.weighted_random_sampling(
+                    qnodes, coeffs, self.max_shots, self.trainable_args, *args, **kwargs
+                )
+            else:
+                device.shots = [(1, self.max_shots)]
+                grads = [
+                    qml.jacobian(expval_cost, argnum=i)(*args, **kwargs)
+                    for i in self.trainable_args
+                ]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def _single_shot_qnode_gradients(self, qnode, args, kwargs):
+        """Compute the single shot gradients of a QNode."""
+        device = qnode.device
+
+        self.check_device(qnode.device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(1)
+
+        try:
+            device.shots = [(1, self.max_shots)]
+            grads = [qml.jacobian(qnode, argnum=i)(*args, **kwargs) for i in self.trainable_args]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def compute_grad(
+        self, objective_fn, args, kwargs
+    ):  # pylint: disable=signature-differs,arguments-differ
+        r"""Compute gradient of the objective function, as well as the variance of the gradient,
+        at the given point.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            args: arguments to the objective function
+            kwargs: keyword arguments to the objective function
+
+        Returns:
+            tuple[array[float], array[float]]: a tuple of NumPy arrays containing the gradient
+            :math:`\nabla f(x^{(t)})` and the variance of the gradient.
+        """
+        if isinstance(objective_fn, qml.ExpvalCost):
+            grads = self._single_shot_expval_gradients(objective_fn, args, kwargs)
+        elif isinstance(objective_fn, qml.tape.QNode) or hasattr(objective_fn, "device"):

```

Comment:
Do we really want to allow anything that isn't a QNode but still has a device attribute? The error below and docs seem to imply that we don't (and I'm not sure if it's an extra benefit if we do)?

Reply:
I put this in as a hidden/advanced feature, because this came up with the metric tensor function, and that was the workaround we went with to allow users to optimize 'illegal' cost functions if they were happy to take the risk!

Just adding it in now in case it needs to be used later in forum/slack questions.